### ISSUE TITLE
help or on handling WebRequest 504

### ISSUE THREAD

**Issue by expand** [App-vNext/Polly#272]
Hi

What is the appropriate refactor to utilize polly for this method? 

```
        private string Query(string address){
            HttpWebRequest webRequest = (HttpWebRequest) WebRequest.Create(new Uri(address));

            try{
                using (WebResponse webResponse = webRequest.GetResponse()){
                    Stream str = webResponse.GetResponseStream();
                    Debug.Assert(str != null, "str != null");
                    using (StreamReader sr = new StreamReader(str))
                        return sr.ReadToEnd();
                }
            }
            catch (WebException wex){
                using (HttpWebResponse response = (HttpWebResponse) wex.Response){
                    if (response == null)
                        throw;

                    Stream str = response.GetResponseStream();
                    using (StreamReader sr = new StreamReader(str)){
                        if (response.StatusCode != HttpStatusCode.InternalServerError){
                            throw;
                        }
                        return sr.ReadToEnd();
                    }
                }
            }
        }

```

**Comment by reisenberger** [comment]
@expand, The question is perhaps a little too broad, to be able to comment usefully.

Have you decided what resilience strategy/ies you wish to use, out of the [available Polly offerings](https://github.com/App-vNext/Polly#resilience-policies)? In introducing Polly into that method, what you are seeking to achieve?  What refactoring have you tried? 

If there a specific refactoring problem you have encountered (for example: "I would like this aspect of the original code to be invoked, when X happens in the real world / or when the Polly policy does Y ... but can't see how to do that with Polly"), then we can probably help!

**Comment by reisenberger** [comment]
@expand What refactor is appropriate would depend on your goal: introduce retries? circuit breaker? or just reproduce what is there already using Polly?

If your goal is to reproduce the existing functionality using Polly:

(_Aside: until enhancement #278 is delivered, it is not actually possible to deliver this refactor.  
We need enhancement #278 to allow the fallback value to be based on the handled exception._)

It looks as if the intent of the existing method is to return the body of the response as a string. Or if `HttpStatusCode.InternalServerError` is returned, to return still the body of the response, if there is a body.  

We could start by factoring out some common code:

```c#
private string GetStreamAsStringFromResponse(HttpWebResponse response)
{
    using (StreamReader sr = new StreamReader(response.GetResponseStream())) return sr.ReadToEnd();
}
```

(I've intentionally elided many of the guard conditions against `null` for brevity and to focus on the structure; you probably want to keep them.)

The exception conditions actually handled by the `catch` in the original code can be condensed into one line, as a handle condition for a Polly policy:

`Policy.Handle<WebException>(wex => wex.Response?.StatusCode == HttpStatusCode.InternalServerError)`

The original code substitutes one thing for another, on error, in two possible ways: one `HttpWebResponse` for another; and ultimately one `string` return value for another.  The Polly policy for providing a substitute response when something fails is [`FallbackPolicy`](https://github.com/App-vNext/Polly/wiki/Fallback).  

`FallbackPolicy` is strongly-typed to the type of result you want to substitute, for [compile-time type-binding with the type of delegates you execute through it](https://github.com/App-vNext/Polly/wiki/Non-generic-and-generic-policies#binding-handleresulttresult-and-executetresult).  So we need a separate `FallbackPolicy` for each of the `HttpWebResponse` and `string` responses.  But we can write the code to express the handled conditions only once:

```c#
var handleConditions = Policy.Handle<WebException>(wex => wex.Response?.StatusCode == HttpStatusCode.InternalServerError);

// NOTE: This overload of FallbackPolicy will not be available until enhancement #278 is delivered.
FallbackPolicy responseFallback = handleConditions 
    .Fallback<HttpWebResponse>((failure, context, cancellationToken) => failure.Exception.Response); // equivalent to mapping to wex.Response in your original catch (WebException wex)

// NOTE: This overload of FallbackPolicy will not be available until enhancement #278 is delivered.
FallbackPolicy bodyFallback = handleConditions
    .Fallback<string>((failure, context, cancellationToken) => { 
         using (HttpWebResponse response = (HttpWebResponse) failure.Exception.Response)
         { 
             return GetStreamAsStringFromResponse(response));
         } 
     });
```

The overall execution then condenses to just a few lines:

```c#
private string Query(string address) {
    HttpWebRequest webRequest = (HttpWebRequest) WebRequest.Create(new Uri(address));
    using (HttpWebResponse response = responseFallback.Execute(() => webRequest.GetResponse()))  
    {
        return bodyFallback.Execute(() => GetStreamAsStringFromResponse(response));
    }

```

Does this help?  Is this the kind of refactoring you were looking for?

**Comment by apobekiaris** [comment]
I posted with wrong account before but yes WOW for Polly what an amazing framework! I guess I have to wait a bit before testing your answer but many thanks in advance!

**Comment by reisenberger** [comment]
@apobekiaris Cool.  Let us know if you have more ideas for Polly or anything we can help with.

**Comment by ExtRemo75** [comment]
See Pull request #303, ready for review

**Comment by reisenberger** [comment]
@apobekiaris  Thanks to @ExtRemo75 's awesome contribution ( :+1: @ExtRemo75 !), the new syntax mentioned [in my original reply](https://github.com/App-vNext/Polly/issues/272#issuecomment-316823864) will be available in Polly v5.4.0 (hopefully release in the next few days)
